src/Compilers/VisualBasic/Portable/Compilation/VisualBasicDeterministicKeyBuilder.vb

```diff
@@ -0,0 +1,109 @@
+        End Sub
+
+        Protected Overrides Sub WriteParseOptionsCore(writer As JsonWriter, parseOptions As ParseOptions)
+            ' This can happen for SyntaxTree that are constructed via the API. 
+            If parseOptions Is Nothing Then
+                Return
+            End If
+
+            Dim basicOptions = TryCast(parseOptions, VisualBasicParseOptions)
+            If basicOptions Is Nothing Then
+                Throw New InvalidOperationException()
+            End If
+
+            MyBase.WriteParseOptionsCore(writer, parseOptions)
+
+            writer.Write("languageVersion", basicOptions.LanguageVersion)
+            writer.Write("specifiedLanguageVersion", basicOptions.SpecifiedLanguageVersion)
+
+            writer.WriteKey("preprocessorSymbols")
+            writer.WriteObjectStart()
+            For Each pair In basicOptions.PreprocessorSymbols.OrderBy(Function(x, y) StringComparer.Ordinal.Compare(x.Key, y.Key))
+                Dim value = pair.Value
+                If value Is Nothing Then
+                    writer.WriteNull(pair.Key)
+                    Continue For
+                End If
+
+                Dim type = value.GetType()
+                If type = GetType(String) Then
+                    writer.Write(pair.Key, CType(value, String))
+                ElseIf type = GetType(Boolean) Then
+                    writer.Write(pair.Key, CType(value, Boolean))
+                ElseIf type = GetType(DateTime) Then
+                    writer.Write(pair.Key, CType(value, DateTime).ToString("G"))
+                ElseIf type = GetType(Char) Then
+                    writer.Write(pair.Key, CType(value, Char).ToString())
+                ElseIf type = GetType(Int16) Then
+                    writer.Write(pair.Key, CType(value, Int16).ToString("G"))
+                ElseIf type = GetType(Int32) Then
+                    writer.Write(pair.Key, CType(value, Int32).ToString("G"))
+                ElseIf type = GetType(Int64) Then
+                    writer.Write(pair.Key, CType(value, Int64).ToString("G"))
+                ElseIf type = GetType(UInt16) Then
+                    writer.Write(pair.Key, CType(value, UInt16).ToString("G"))
+                ElseIf type = GetType(UInt32) Then
+                    writer.Write(pair.Key, CType(value, UInt32).ToString("G"))
+                ElseIf type = GetType(UInt64) Then
+                    writer.Write(pair.Key, CType(value, UInt64).ToString("G"))
+                ElseIf type = GetType(Decimal) Then
+                    writer.Write(pair.Key, CType(value, Decimal).ToString("G"))
+                ElseIf type = GetType(Single) Then
+                    writer.Write(pair.Key, CType(value, Single).ToString("G"))
+                ElseIf type = GetType(Double) Then
+                    writer.Write(pair.Key, CType(value, Double).ToString("G"))
+                ElseIf type = GetType(SByte) Then
+                    writer.Write(pair.Key, CType(value, SByte).ToString("G"))
```

Comment:
:'(

Reply:
Yeah I couldn't see a better way to do this which was durable across cultures (once I switch `G` to invariant)

Reply:
Doh, I can just use `IFormattable` and simplify this a lot